### ISSUE TITLE
Indexing within a partition is inefficient

### ISSUE THREAD

**Issue by avikivity** [scylladb/scylla#4185]
Consider

```
    CREATE TABLE tab (
        pk text,
        ck text,
        val text,
        PRIMARY KEY (pk, ck)
    );

    CREATE INDEX idx ON tab(val);
```

This will work well for the global query

    SELECT * FROM tab WHERE val = ?

but very poorly on the local query

    SELECT * FROM tab WHERE val = ? AND pk = ?

because the index intermingles rows from all partitions that have the queried `val`, can cause imbalance, and generates extra hops.

We could improve the situation by providing syntax that instantiates the index as materialized view with a primary key of (pk, val, ck). Then the select would transform to a slice of a partition that is co-located with the data.

**Comment by duarten** [comment]
Maybe we can simply warn on these types of queries when they are prepared, and suggest the user employ MVs. This generalizes better than adding syntax for particular cases (a very narrow one in this case, since range queries on pk wouldn't work without filtering).

**Comment by avikivity** [comment]
It's not that narrow, since it matches the case where Cassandra works well. It's easier to migrate if you sprinkle some sugar on the CREATE INDEX statement vs. your queries.

**Comment by nyh** [comment]
There's another way to look at Avi's idea: The query Avi wrote was
```
SELECT * FROM tab WHERE val = ? AND pk = ?
```
If we look at it the other way around:
```
SELECT * FROM tab WHERE pk = ? AND val = ?
```
You can think about this as a "local index" feature inside every very long partition, meaning that val=? is searchable inside every partition, which sounds like a useful and non-Scylla-specific need.

We could invent a new syntax for this, or use the existing (and ugly) "custom index" syntax for this (with the assumption that one non-partition-key column is indexed, together with the entire unmodified partition key.

**Comment by denesb** [comment]
I'm not very familiar with how this works, but can't we just fall back to filtering the single partition in this case? I think we should try to move into the direction of making CQL simpler to use, instead of moving even more expert decisions to them.

**Comment by duarten** [comment]
@denesb In that case you would need to scan entire, possibly huge partitions. That would defeat the intent of the user in indexing the column.

**Comment by avikivity** [comment]
I think we need syntax for both global and local indexes, so you explicitly say what queries you want to optimize for. The non-specific syntax should emit a warning.

**Comment by slivne** [comment]
how would we emit the warning ?

At times users have very little control/knowledge over SI implementation - they are using a third party (Kong, Titan, Karios etc.) that may use SI and is supposed to work out of the box. The schema may be in the third party tool code - are they supposed to change the code to get it working, I think thats not reasonable.

We should make this work well out of the box.

Since Cassandra's default is local we should probably change ours to be the same. 

Changing the default to align with Cassandra has a draw back, Cassandra  allows on its local indexes global queries (in a non efficient manner) - preferably we would like this to work out of the box as well. The solution to that is to add a Global index as well.

While adding both Local+Global indexes entails a performance hit on write path and space overhead it seems to me to be the "lesser of three evils", than to have the user:
-  fail on global queries (in case SI is defaulted only Local index) and trying to figure that out
 - perform very badly on local queries (in case SI is defaulted only to Global index like it is today).

**Comment by slivne** [comment]
@tzach / @avikivity  - I want to try and push this into 3.1

**Comment by duarten** [comment]
I still think we can squeeze more performance out of the global indexes still, so the situation may not be so dire.

Regarding third-party tools, maybe we can provide nodetool commands to change the index type, which would at least provide a workaround.

Maybe we can even use the per-table partitioning scheme to help co-locate the indexes and the base tables. In the future.

**Comment by avikivity** [comment]
Usually, the user needs just one, and I don't think we can decide in advance which it is. So we should just let the user decide.

**Comment by duarten** [comment]
That could be solved by an approach in the spirit of #3810, where we can decide to query the base table (potentially by using filtering) or the index. Still not the best performance, but also not the worst.

**Comment by avikivity** [comment]
I think we should aim for the simple cases first, where the user knows the queries up-front and can therefore specify the correct index type.

**Comment by slivne** [comment]
@avikivity I don't understand you solution for a user installing a third party that is using SI - what will they get ?

**Comment by avikivity** [comment]
We can have a configuration item that says what's the default secondary index type.

**Comment by duarten** [comment]
I agree. Another approach is yet to have special syntax for indexes to be created based on a query. Each query would be backed by a different secondary index, unless we could efficiently support multiple queries on the same index. Since with this approach we know the queries upfront, we can come up with the best layout.

**Comment by slivne** [comment]
and what is the default configuration local/global ?

in my view local

**Comment by slivne** [comment]
```
CREATE TABLE tab ( 
  pk text, 
  ck text, 
  val text, 
  PRIMARY KEY (pk, ck) ); 
  CREATE INDEX idx ON tab(val); 
  CREATE GLOBAL INDEX idx_global(val);
  CREATE SECONDARY INDEX idx_local(val);
```

- CQL ```CREATE GLOBAL INDEX```
   - will create a global index (like we do in scylla today) - an index with a backing materialized view with (val,pk,ck)
   - queries ```select .... where val=Y``` will be supported
   - queries ```select ... where pk=X and val=Y``` will return an error "Cannot execute this query as only a global index is defined and thus the query may have unpredictable performance. If you want to execute this query create a local index or force running it despite the performance unpredictability, using ALLOW FILTERING"

- CQL ```CREATE LOCAL INDEX```
   - will create a local index - an index with a backing materialized view with (pk,val, ck)
   - queries ```select ... where pk=X and val=Y``` will be supported
   - queries ```select .... where val=Y``` will return an error "Cannot execute this query as only a local index is defined and thus the query may have unpredictable performance. If you want to execute this query create a global index or force running it despite the performance unpredictability, using ALLOW FILTERING" 
  
- CQL ```CREATE INDEX```
   - will create an index based on the secondary_index_default_type config value
   - on creation the type will be decided and the type will be stored in system_schema.indexes - so that altering the default will not affect created indexes

configuration
```
# secondary_index_default_type sets the type of index set in case the user does not specify it explicitly in create index. possible values:
# local - a local index
# global - a global index
# secondary_index_default_type : local 
```

**Comment by slivne** [comment]
@avikivity / @duarten  - please review comment - so we can assign and implement

**Comment by avikivity** [comment]
I prefer WITH clauses, they are more extensible.

**Comment by nyh** [comment]
@slivne my comments:
1. your "GLOBAL" / "LOCAL" syntax is nice, but arguably Cassandra already has a (much uglier) syntax for custom index types - which they used for SASI, we could also reuse that if we don't want to add new syntax. But maybe it's time to break away, and add new syntax!
2. When you create a LOCAL index and search on just val=Y, it's worse than "unpredicatable" performance - you'll actually (unless I'm missing something?) need to scan the entire table, so it's predictably bad. So the message should be even more apprehensive.

**Comment by duarten** [comment]
I think we should give some thought to query-driven index creation, something like

`CREATE INDEX [name] ON tab(col) WHERE restriction [AND restriction]*`

Or something.

**Comment by avikivity** [comment]
That's very unorthodox. Let's try to copy existing SQL syntax.

**Comment by tzach** [comment]
> CQL `CREATE INDEX`
> 
> * will create an index based on the secondary_index_default_type config value
> * on creation the type will be decided and the type will be stored in system_schema.indexes - so that altering the default will not affect created indexes

This second level of configuration is redundant and confusing.
Why should the user have a reasonable config for all his different workload? Its better to just have the default as a local index, and not add a confusing configuration
Local make more sense as default as it is more similar to Apache Cassandra SI

**Comment by tzach** [comment]
> CQL `CREATE GLOBAL INDEX`

CQL support the following syntax
```
create_index_statement ::=  CREATE [ CUSTOM ] INDEX [ IF NOT EXISTS ] [ index_name ]
                                ON table_name '(' index_identifier ')'
                                [ USING string [ WITH OPTIONS = map_literal ] ]
index_identifier       ::=  column_name
                           | ( KEYS | VALUES | ENTRIES | FULL ) '(' column_name ')'
```
source: http://cassandra.apache.org/doc/latest/cql/indexes.html#create-index

So a more natrual syntax will be
```
CREATE INDEX ON users (email) WITH OPTIONS = {'type': 'global'};
```

**Comment by slivne** [comment]
@avikivity 

the drawback of using WITH is that its usually something that can be modified via ALTER ... and I do not think we want to support switching from global to local - it seems as an overkill - the user should delete and create an index if that is needed.

with regards to @tzach suggestion

thats not possible ... there is a limitation of OPTIONS only for CUSTOM index

```
cqlsh > create INDEX idx on keyspace1.standard1 ("C0") WITH OPTIONS = {'scylla_type': 'global'};

InvalidRequest: Error from server: code=2200 [Invalid query] message="Cannot specify options for a non-CUSTOM index"
```

we can remove this limitation yet maybe as @nyh suggested it actually makes it a bit more clear to use custom indexes.

So if we want anything that is similar to cassandra allowed syntax its 

```
create custom index XXX on YYY (...) USING "scylla.global.index"
create custom index XXX on YYY (...) USING "scylla.local.index"
````

If we will need additional arguments we can add them via the ```WITH options```

**Comment by slivne** [comment]
@avikivity / @duarten - please comment

**Comment by avikivity** [comment]
I think we can make the syntax more flexible and allow WITH without using USING.

**Comment by slivne** [comment]
@avikivity - their is drawback with WITH, are you expecting the solution to support moving from global to local indexes ? Users are "taught" that WITH attributes can be modified with ALTER, I think it would be preferable to maintain this and not add a limitation.

If you still want this kept with WITH - are you ok with tzach's suggestion above ?

**Comment by avikivity** [comment]
I don't think everyone expects every WITH option to be ALTERable.

But, `CREATE GLOBAL INDEX` and `CREATE LOCAL INDEX` are also good options. Need to see what modern SQL dialects do (they have many index types).

**Comment by avikivity** [comment]
> 2\. When you create a LOCAL index and search on just val=Y, it's worse than "unpredicatable" performance - you'll actually (unless I'm missing something?) need to scan the entire table, so it's predictably bad. So the message should be even more apprehensive.

We should just reject global searches on local indexes. The user can create a global index if they're interested in that.

Alternatively, we can try to implement node-by-node scans, but I think it's counterproductive. You add an index in order to be efficient, not in order to find diffferent ways to be inefficient.

**Comment by avikivity** [comment]
Here's Transact-SQL syntax (SQL Server):

```
CREATE [ UNIQUE ] [ CLUSTERED | NONCLUSTERED ] INDEX index_name   
    ON <object> ( column [ ASC | DESC ] [ ,...n ] )   
    [ INCLUDE ( column_name [ ,...n ] ) ]  
    [ WHERE <filter_predicate> ]  
    [ WITH ( <relational_index_option> [ ,...n ] ) ]  
    [ ON { partition_scheme_name ( column_name )   
         | filegroup_name   
         | default   
         }  
    ]  
    [ FILESTREAM_ON { filestream_filegroup_name | partition_scheme_name | "NULL" } ]  
  ```

So it looks like inserting adjectives between `CREATE` and `INDEX` is traditional and expected.

**Comment by avikivity** [comment]
btw, `INCLUDE ( column_name, ...)` adds more columns like materialized views, we could support that syntax easily.

**Comment by slivne** [comment]
So @avikivity what should it be ?

```CREATE LOCAL INDEX / CREATE GLOBAL INDEX``` ?

**Comment by avikivity** [comment]
Yes (and CREATE INDEX works as it does now, creating a global index).

**Comment by slivne** [comment]
@eliransin - please assign and lets agree on the complete solution - in which system_schema tables do we save the global vs local indication etc.

**Comment by psarna** [comment]
I can volunteer for this one, and most of the work can be done even before we decide in which system_schema tables we save global vs local, etc. So if there are no objections, somebody feel free to assign it to me.

**Comment by dorlaor** [comment]
@psarna look like you were randomly selected :)

**Comment by avikivity** [comment]
Alternative syntax idea:

`CREATE INDEX idx1 ON tab(some_column)` = global index
`CREATE INDEX idx2 ON tab(partition_key, some_column)` = local index

This fits with how indexes are traditionally created. For composite partition keys,

`CREATE INDEX idx2 ON tab((pk1, pk2, pk3), some_column)` = local index, clear that no sorting internal to the partition key is performed since it's treated as a tuple

**Comment by psarna** [comment]
Fixed with a9cf07369ff8dc2c88b0538dc5052d49d4f3204e
